### DISCUSSION TITLE
Mutable compound shapes and deterministic roll back

### DISCUSSION THREAD

**Discussion by Ivan-le-sketch** [jrouwe/JoltPhysics#1391] [Q&A] [answered]
The Being Sloppy While Still Being Deterministic section of the documentation mentions :
> the main thing you need to ensure is that when you do a [PhysicsSystem::Update](https://jrouwe.github.io/JoltPhysics/class_physics_system.html#a4cdc76b2486dfaecfc9319d040e5e5f3) that the binary state is the same

What would be the best way to handle mutable compound shapes modifications when rolling back to a previous simulation state ?
There does not seem to be a way to restore the subshape index when trying to revert a subshape addition/removal.
A very naive approach would be to keep a recent history of the compound shape by cloning it for every save point but I'm not sure if there is a better way.

**Comment by jrouwe** [comment, marked as answer]
Keeping a history is currently indeed the only way you can restore shapes to their previous state. If this is impractical, I can add a method that allows inserting a shape in the middle.

**Reply by Ivan-le-sketch** [reply]
Creating a clone per save point does not sound too impractical but I may benefit from having the possibility to insert a shape with a specified index. I would have to test if the cost for cloning a compound shape exceeds the cost for rebuilding one for different scenarios.

**Reply by jrouwe** [reply]
I have to say that we never modify a mutable compound shape, we always create a new one. We do this because it makes it multi thread safe in case one thread is querying the shape while another is changing it.

**Comment by jrouwe** [comment]
I added the ability to add a sub shape at a particular index in #1414
